### DISCUSSION TITLE
How to make a question/answer system in tock ?

### DISCUSSION THREAD

**Discussion by dcr31000** [theopenconversationkit/tock#1381] [Q&A]
Hello,
I'm looking to make a system to ask the user for information (message, contact details) to send an email afterwards. 
For this, I build my bot with the API (node). 

But I don't understand how to request information from the user? (In the idea, do a bot.send("Your message?"), wait for its response and continue with bot.send("Your email?") etc.).

Thank you in advance for your help 🙂

**Comment by francoisno** [comment]
Hello @dcr31000,

A story is triggered for every sentence of the user, yet it is possible to write a story checking if the information is in context, otherwise asking for it. Next sentence, provided more information has been added to context, the story triggered again will check again and follow another path, doing something with the information.

Here a basic example of such a story, in _Bot API_ mode:

                newStory("form") {

                    send("Let's do something with a name and date...")

                    val name = entityText("name")
                    val birthday = entityValue<DateEntityValue>("birthday")?.date

                    if (name == null) {
                        end("What is your name?")
                    } else {
                        if (birthday == null) {
                            end("When is your birthday?")
                        } else {
                            // TODO: Do something interesting with the name and date?
                            send("I have everything I need...")
                            end("...I know your birthday is $birthday, thanks $name!")
                        }
                    }
                }

The code would be very similar in _Bot Integrated_ mode.

The trickiest part is generally not to implement the logic in the story, but to build the intent/NLP model so that the story is triggered again once the information is given. In _Bot Integrated_ mode, it is possible to apply constraint on the next sentence intent detection, which makes it really easier to implement funnels. _Secondary intents_ are also a mean to implement Q/A in stories.

An example in _Bot Integrated_ mode is provided in the `AskIdentityHandler.kt` from the Bot Open Data sample: https://github.com/theopenconversationkit/tock-bot-open-data/blob/master/src/main/kotlin/story/AskIdentityHandler.kt

    story("ask_identity", secondaryIntents = setOf(cancel)) {

        if (isIntent(cancel)) {
            end("Ok by")
        } else {

            //we have not detected entities - ok so we assume the whole test is the entity missing
            if (!hasActionEntity(firstNameEntity) && !hasActionEntity(lastNameEntity)) {
                if (firstName != null && lastName == null) {
                    lastName = userText
                } else {
                    firstName = userText
                }
            }

            //only ask_identity intent or cancel intent allowed for next request
            nextUserActionState = NextUserActionState(
                askIdentity to 0.4,
                cancel to 0.0
            )
            when {
                firstName == null -> end("What is your first name?")
                lastName == null -> end("Ok {0}, what is your last name?", firstName)
                else -> end("Your complete name is {0} {1}. Do you want to change it or quit?", firstName, lastName)
            }
        }

    }

Hth. Please let me know! :)
